Team Data Optimizations

harishbommakanti opened the pull request:
- simplified method of match/pit data storing (array of len 1 objects to just an object)
- added custom notes section for match data
- changed team radio list to select
- changed table to w3-grid for pit scouting, inputs to textareas for pit scouting

harishbommakanti committed:
Create CNAME file

harishbommakanti committed:
Merge branch 'master' into gh-pages

harishbommakanti committed:
Minor change in the nav list html code

- home.html --> index.html

harishbommakanti committed:
Merge branch 'master' into gh-pages

harishbommakanti committed:
Merge branch 'master' into gh-pages

harishbommakanti committed:
Merge branch 'master' into gh-pages

harishbommakanti committed:
Initial Commit: Set up Jekyll environment

- installed rubyinstaller/ruby/gem/jekyll and set up all the jekyll environment

harishbommakanti committed:
Created gitignore, removed jekyll-cache and _site/ files

- gitignore updated to ignore cache and site output files
- removed cache and site output files

harishbommakanti committed:
Created nav.html file

- added nav.html to file, will start to implement it in the pages

harishbommakanti committed:
Used jekyll include for nav button

-  added nav button html code to nav.html and used liquid include tags to reference it

harishbommakanti committed:
Updated nav.html: added liquid comment

harishbommakanti committed:
Created head.html

- copied all of the css/js/font sheets to an includes head.html file

harishbommakanti committed:
Used jekyll include head.html on all pages

- used jekyll include head.html on all pages
- fixed a minor bug for button styling

harishbommakanti committed:
Created a basic layout file/directory

harishbommakanti committed:
Added code to the basic.html layout

- has the template for the basic layout on all pages
- modified season page to make its body tag similar to others (instead of onchange, I just added that to the doc.onReady in the js file)

harishbommakanti committed:
Used basic.html as a layout on all pages

- used basic.html as a layout on all html pages to avoid copying over code ie increasing possibility of buggy code

harishbommakanti committed:
minor fix: deleting commented out code

- deleted old commented out html code

harishbommakanti committed:
PR fix: updated gitignore file

will change the navbar to be better on mobile devices

harishbommakanti committed:
Added css/html for a navbar

- used w3css and my own styling to go from a navbar button to an actual navbar
- deleted navbar button code (js/html/css)
- will now work on creating an active tab

harishbommakanti committed:
Added more styling to navbar

harishbommakanti committed:
PR fix: re-added links to different pages on index.html

harishbommakanti committed:
PR fix: updated styles.css for a more efficient way of making all elements ubuntu maroon color

harishbommakanti committed:
Created pages.yaml, added _data folder

- added file to make the collapsable navbar on smaller screens easier to use with jekyll templating

harishbommakanti committed:
Navbar more friendly to smaller screens

- now the navbar is collapsable on smaller screens

harishbommakanti committed:
PR fix: moved title to navbar

- moved title as part of the red navbar chunk as it looks a little better

harishbommakanti committed:
PR fix: removed commented code in basic.html

harishbommakanti committed:
PR fix: simplified html of navbar

- moved the title into the navbar div

harishbommakanti committed:
PR fix: meta tag edited, some navbar styling

- meta tag edited so the site renders well on the chrome editor

harishbommakanti committed:
PR fix: title and menu are on same line

- added an extra title div that adjusts so its on the same line with the menu button, makes all the home/season config... buttons centered while not making the navbar look too bad

harishbommakanti committed:
Removed github/instagram/twitter links from main page

- removed html/css for social media links on front page as they are 1) not really linked to frc scouting and 2) kind of hard to style across screen sizes

harishbommakanti committed:
Further title styling, removed some uneccesary text on index.html

harishbommakanti committed:
PR fix: separated meta tags

harishbommakanti committed:
PR fix: removed unecessary div tag from team-data.html

harishbommakanti committed:
Initial Commit of Team Data Opmtimizations PR: simplified pit q/a storing method

- simplified way of storing/showing pit questions/answers (look into the github project card for more info)

harishbommakanti committed:
Match data storing/showing is also simpler

- array of len 1 dictionaries --> 1 dictionary of multiple key/value pairs

harishbommakanti committed:
Updated comments on importing/exportin

- due to the last 2 changes (array of dictionaries to a straightforward dictionary) I'm just adjusting the comments so that ppl won't be confused in the future

harishbommakanti committed:
Added a custom notes section for team data

- added a notes section for the person to include info like 'tippy' or 'doesn't cooperate with teamates well' to be defining factors during gameplay

harishbommakanti committed:
Changed radio list to select

- radio list to select as events will have upwards of 20-30 teams, don't want to have to scroll through all of that

harishbommakanti committed:
Updated loadPit() for new pit method

- a bit late of a commit, but I forgot that this was a thing...
- loading previous answers for the pit input boxes now works with the simpler storage method

harishbommakanti committed:
Fixing bug with loading previous pit answers

- fixed basically a nullpointer exception

harishbommakanti committed:
Changed th pit table to w3-grid

- changed the pit table to w3-grid to be better formatted on mobile
- needs further styling and I need to revise code for reading the table since the layout changed

harishbommakanti committed:
Revised savePit() and loadPit() to make sure they worked with textarea and w3-grid

- revised other loading/saving methods to make sure they read from textareas instead of inputs, and w3-rows instead of tables

harishbommakanti committed:
Added autosize.js from Jack Moore's github profile

harishbommakanti committed:
Finished using autosize.js forthe pit textareas

- realized that I needed the whole github project and not just the autosize.js, put the autosize-master under the scripts folder
- used the autosize method in loadpit() and it works fine

harishbommakanti committed:
Merge branch 'master' into Optimizations

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -317,6 +315,9 @@ function loadMatch()
         
         $('#matchQuestionTable').append(newTRHtml);   
     }
+    var customNotesInputHTML = `<input type="text" class="match-frequency js_clear_on_load">`; //for custom notes each match
```
good idea - but don't require this to save an entry

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -434,9 +434,8 @@ function sortMatchOutputTable(savedLocStrKeys) //sorts the table by team and the
 
             for (var k = 0; k < innerKeys.length; k++) //displaying answers to each question
             {
```
so, if the note that is added is really-long - it has the same scrolling in the input problem as the questions did (good job utilizing the auto-size - it works really well).

But you'll also run into an issue with it forcing the page width really wide. In theory, you could fix this by imposing a max-width, but it would probably be simpler to swap this for a w3-grid as well - except for the fact that you don't know how many items you have at coding time. Think about this and let me know what you come up with.

harishbommakanti replied:
I was tinkering with it and all of the classes added on to the regular table (w3-responsive, w3-table, w3-container etc) didn't have an effect on this problem. I also looked at the w3-grid as well, but you're right, we don't know how many questions the user has and therefore can't assign the divs a class like m6 beforehand.

If it's possible, maybe we can generate classes based on how many questions there are. Like, for the pit table, if there are 6 questions, added on to the column for the team #, the class could be .w3-col.sevenths, with the width being 100/7 % for each column. Or there could be a class called equalWidth and every time the user wants to display the page, the width in the class could be changed based on the # of questions, so the table's width would be adjusted each time the showAllPitTable() method is called (same with the match data table, but probably a different class for that).

If this is not realistic, I'm thinking that maybe we just impose a max-width element in the table rows, but I'm unsure of how to do that because width:auto; hasn't really been working for me. And if we try to set a specified # of pixels as the max width, that would vary based on screen size.

mshafer1 replied:
Some more thoughts to throw out there (to make sure we've considered at least most of the reasonable options)

What if we always set everything to s2 (or s(12/len(questions) - whichever is larger) and let it wrap on it's own?
(so, if there were more than 6 questions/items, it would aut-wrap at the sixth, but keep every cell 1/6th - or whatever - of the screen wide). Would it look bad if one of the those had a really long text in it?

What if we turned the table? (using set column widths, but allowing it to scroll to the side)

|Team #|3288|3937|6321|21017|
|---|---|---|---|---|
|Match #|1|12|23|45|
|Balls|2|5|1|3|
|Hatch|1|3|8|2|
|Comments|Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque vel |Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque vel |Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque vel |Lorem ipsum |dolor sit amet, consectetur adipiscing elit. Pellentesque vel |

What if we made it only show the current team and put it where match #'s were "header" rows and everything but comments shared an even space and comments got their own row?

|Balls|Hatch|
|---|---|
|Team: 3288||
|Match: 1||
|1|2|
|Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque vel||
|Match: 9||
|3|4|
|Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque vel||
|Match: 15||
|0|8|
|Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque vel||

(might need to apply some padding to make it line up in a skim-able way)

harishbommakanti replied:
So I tried to replace the pit table with the w3-grid table but the problem is that continuous strings like long words/numbers aren't wrapping within the divs. I also used the s[12/len(questions)], that was a good tip.

Here's a pic of the table:
![image](https://user-images.githubusercontent.com/33380236/72101223-d5906400-32e9-11ea-8eb4-dccb07e3c83f.png)

and when it's on mobile:
![image](https://user-images.githubusercontent.com/33380236/72101270-f1940580-32e9-11ea-8ac0-778976c9a6e9.png)

I added table borders and long words like teammates and autonomous which extended past the borders of the divs, interfering with other divs and making the width of the responsive table enlarged. So I think the problem is long strings extending past the div borders.

harishbommakanti replied:
Wait, I solved that problem with `overflow-wrap: break-word`. Here is how it is now:
![image](https://user-images.githubusercontent.com/33380236/72103745-77ff1600-32ef-11ea-9d33-a061cf58b5bc.png)

Looking at this, I think we might need to change the layout, like you were saying. Seeing that the questions in the pic are probably reasonable length, I think the best solution would be to have the left-most column being Team# and the questions (Team#, question 1...) and each following column having data for each team.
So:
Team # | Team 1 | Team 2 | Team 3
q1         | yes       | no        | yes
...

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -135,74 +135,76 @@ function loadPit()
+
+        //only put a top border on the first question row for styling: making sure top/bottom borders dont overlap for non-first rows
+        var newTRHtml = (i==0)?`<div class='${w3RowClassText}'> ${insideHtml} </div>`:
+                                `<div class='${w3RowClassText} tableBorders-NoTop'> ${insideHtml} </div>`;
```
You can use css to give you this.

This line becomes:

```js
var newTRHtml = `<div class='${w3RowClassText} tableRowBorders-NoTop'> ${insideHtml} </div>`
```

And add to your css:
```css
table .tableRowBorders-NoTop:not(:first-child) {
  border-top:none;
}
```

Where, now you able to say for all elements in a table with the class `tableRowBorders-NoTop` - except the first one, apply this rule.

mshafer1 reviewed (approved):
works - but you might want to think about what happens if the user puts in a long comment (see note).

(also, checkout out ccs selectors: https://www.w3schools.com/cssref/css_selectors.asp, they let you keep the styling in the css and the contents in the js - without logic of when to apply a class or not)

harishbommakanti committed:
PR fix: simplified newTRHtml code with css

harishbommakanti committed:
PR fix: Custom Notes not required to save match answers

harishbommakanti committed:
Pit table changed to w3-grid

- removed code for having the allTeamsPitTable being a table
- used a w3-grid
- added some css

harishbommakanti committed:
added a css class to wrap the div contents within the div

harishbommakanti committed:
PR fix: Interim switch bewteen w3-grid layouts

- change to header on top to header on left for the w3-grid for the allTeamsPitAnswers table
- still need to normalize heights/widths and add a scrollbar, hence the interim part

harishbommakanti committed:
PR fix: w3-grid layout fixed: questions are on the left side

- only fix to be done on this part now is to have a scrollbar for horizontal scrolling
- added some css for styling

harishbommakanti committed:
PR fix: reverted allPitTable w3-grid to an actual table, but 'rotated' the table left (header on left etc)

- reverted to a table format after having issues with width and standardizing rows
- added a horizontal scrollbar
- will now work on doing the same with the matchDataTable

harishbommakanti committed:
PR fix: cumulative match data table also re styled

- cumulative match data table is also styled and is given a max width: will show a horizontal scroll bar when the # of teams makes the table stretch out too far to the right

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -239,17 +239,17 @@ function showAllPitDataTable() //shows pit stats of all teams for the current ev
-    var tableHeader = '<tr><th class="allTeamsRow">Team #</th>'; //top-left most cell
+    var widthString = `s${parseInt(12/(questionsArr.length+1))}`; //+1 for the Team# column
+    var tableHeader = `<div class="w3-col ${widthString} w3-center"> <h4>Team #</h4> </div>`; //top-left most cell
+    
```
I'm confused, I thought we were switching back **to** `<table>` because it would handle extending horizontally better . . .

harishbommakanti replied:
I think its outdated. Does it look good to merge the PR?

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -213,11 +215,9 @@ function savePitAnswers() //onclick for save pit button
         {
-            all_team_QA_pairs_obj.push(pair);
+            all_team_QA_pairs_obj[question] = answer;
         }
```
flip this if around, if it's the first row, leave early - otherwise handle it:

```js
if (question == 'Question') {
   return;
}

all_team_QA_pairs_obj[question] = answer;
```

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -135,74 +136,75 @@ function loadPit()
-        var inputHTML = `<input type="text" value="${potentialAnswer}" class="answer">`; //pot answer is either '' or what is stored
-        var newTRHtml = `<tr><td class='question scoutingTableRow'>${question}</td><td class='scoutingTableRow'>${inputHTML}</td></tr>`;
+        var classText = 'w3-col l6 w6 s12 question w3-center'; 
+        var inputHTML = `<textarea width='50em' style='resize:vertical' value='${answer}' class='answer'>${answer}</textarea>`; //textarea
```
use `width:80%` (or some high percentage) in the style string this will make the box dynamically almost as big as the area it can fit in (width wise).

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -232,55 +232,71 @@ function savePitAnswers() //onclick for save pit button
-        tableHeader += `<th>${questionsArr[i]}</th>`;
+        var questionWithNoSpaces = stripAlphaNumerics(questionsArr[i]);
+        var questionHeaderHTML = `<th class="allPitTableHeader">${questionsArr[i]}</th>`;
+        var newRowHTML = `<tr class='${questionWithNoSpaces}'>${questionHeaderHTML}</tr>`; //the question is one of the classes (w/o whitespaces), will help with adding team data
```
this is probably a case where we're streaching the class use a little too far.

If you're setting this to easily find the right row to edit, then make it's own attribute.

```js
`<tr question='${questionWithNoSpaces}'>${questionHeaderHTML}</tr>`
```
then to find this row, use 
```js
$(`tr[question='${questionWithNoSpaces}']`)
```
to find it. (this has the benefit of no you can use the question itself, without modifying it to string spaces and such out).

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -135,74 +136,75 @@ function loadPit()
 
-        var inputHTML = `<input type="text" value="${potentialAnswer}" class="answer">`; //pot answer is either '' or what is stored
-        var newTRHtml = `<tr><td class='question scoutingTableRow'>${question}</td><td class='scoutingTableRow'>${inputHTML}</td></tr>`;
+        var classText = 'w3-col l6 w6 s12 question w3-center'; 
```
don't center these - with the questions undoubtedly varying in length, center will make it difficult (subliminally) to  scan to the next line and make the left side look more like a poem and less like it has something to do with the right side.

mshafer1 commented on scripts/teamDataScript.js:
```diff
@@ -232,55 +232,71 @@ function savePitAnswers() //onclick for save pit button
-        tableHeader += `<th>${questionsArr[i]}</th>`;
+        var questionWithNoSpaces = stripAlphaNumerics(questionsArr[i]);
+        var questionHeaderHTML = `<th class="allPitTableHeader">${questionsArr[i]}</th>`;
+        var newRowHTML = `<tr class='${questionWithNoSpaces}'>${questionHeaderHTML}</tr>`; //the question is one of the classes (w/o whitespaces), will help with adding team data
```
this didn't occur to me before, but we should do a few things here:
* sanitize html like code (to prevent what was meant as text from getting rendered)
* replace carriage-return and line feed combinations with `<br/>`  (so that blank lines in the notes stay as blank lines - if someone took the time to put a separator in for readability, we should respect it).

For sanitizing (and do this before you do the new lines), just replace all `<` characters with `&lt;` (so they render as a `<` but don't get interpreted as html code). (`...>${questionHeaderHTML.replace('<', '&lt;')}</tr>`)

For replacing new lines/carriage-returns, add `.replace(/(\r\n|\n|\r)/gm, "<br/>")`
(the `/(\r\n|\n|\r)/gm,` is a regular expression replacement that will match Windows, Linux/Android, and Mac line seperators).

The end effect should be something like:
```js
...>${questionHeaderHTML.replace('<', '&lt;').replace(/(\r\n|\n|\r)/gm, "<br/>")}</tr>
```
(although, you could make it a function).

mshafer1 replied:
('nother though on sanitation) since ampersands are used as html to signify a character, they will get "dropped" if they're typed in (and not a valid unicode value like `&lt;` or `&1F642;`, so replace them to (`.replace('&', '&amp;')`) (make sure to do this first so that you don't turn your less than's into ampersand-lt's).

mshafer1 reviewed (approved):
Couple of comments on coding (and a touch up on handling new lines and such).

The new tables work pretty well. (not gating this PR) you should add a task to the project to make the headers and first column "sticky" so that they are always visible (this makes it so that everything else scroll under them, no scrolling to the top to see which team this note is on, or to the left to check which question I'm looking at).

harishbommakanti committed:
PR fix: minor code fixes

- fixes such as changing textarea width and flipping conditional checking

harishbommakanti committed:
PR fix: edited way to append td's in the table

- instead of abusing class usage I used 'question' and 'element' attributes, allowing me to use the question itself instead of taking the spaces and other nonalphanumerica questions off
- edited small bug for string comparison: == instead of ===

harishbommakanti committed:
PR fix: added method for sanitizing HTML

- added method for sanitizing an HTML line, still need to implement it

harishbommakanti committed:
PR fix: implemented the sanitizeHTMLLine method

- implemented the method in the pit table/match table code
